umpost/um2netcdf.py

```diff
@@ -99,46 +126,193 @@ def convert_proleptic(time):
+                f"Expected {LATITUDE}."
+            )
+
+    if is_lat_river_grid(lat_coordinate.points):
+        lat_coordinate.var_name = VAR_NAME_LAT_RIVER
+    elif is_lat_v_grid(lat_coordinate.points, grid_type, dlat):
+        lat_coordinate.var_name = VAR_NAME_LAT_V
     else:
-        lat.var_name = 'lat'
-
-    lon = cube.coord('longitude')
-    if len(lon.points) == 360:
-        lon.var_name = 'lon_river'
-    elif (lon.points[0] == 0 and grid_type == 'EG') or \
-         (np.allclose(0.5*dlon, lon.points[0]) and grid_type == 'ND'):
-        lon.var_name = 'lon_u'
+        lat_coordinate.var_name = VAR_NAME_LAT_STANDARD
+
+
+def fix_lon_coord_name(lon_coordinate, grid_type, dlon):
+    """
+    Add a 'var_name' attribute to a longitude coordinate object
+    based on the grid it lies on.
+
+    NB - Grid spacing dlon only refers to variables on the main
+    horizontal grids, and not the river grid.
+
+    Parameters
+    ----------
+    lon_coordinate: coordinate object from iris cube (edits in place).
+    grid_type: (string) model horizontal grid type.
+    dlon: (float) zonal spacing between longitude grid points.
+    """
+
+    if lon_coordinate.name() != LONGITUDE:
+        raise ValueError(
+                f"Wrong coordinate {lon_coordinate.name()} supplied. "
+                f"Expected {LATITUDE}."
+            )
+
+    if is_lon_river_grid(lon_coordinate.points):
+        lon_coordinate.var_name = VAR_NAME_LON_RIVER
+    elif is_lon_u_grid(lon_coordinate.points, grid_type, dlon):
+        lon_coordinate.var_name = VAR_NAME_LON_U
     else:
-        lon.var_name = 'lon'
+        lon_coordinate.var_name = VAR_NAME_LON_STANDARD
+
+
+def is_lat_river_grid(latitude_points):
+    """
+    Check whether latitude points are on the river routing grid.
+
+    Parameters
+    ----------
+    latitude_points: (array) 1D array of latitude grid points.
+    """
+    return len(latitude_points) == NUM_LAT_RIVER_GRID_POINTS
+
+
+def is_lon_river_grid(longitude_points):
+    """
+    Check whether longitude points are on the river routing grid.
+
+    Parameters
+    ----------
+    longitude_points: (array) 1D array of longitude grid points.
+    """
+
+    return len(longitude_points) == NUM_LON_RIVER_GRID_POINTS
+
+
+def is_lat_v_grid(latitude_points, grid_type, dlat):
+    """
+    Check whether latitude points are on the lat_v grid.
+
+    Parameters
+    ----------
+    latitude_points: (array) 1D array of latitude grid points.
+    grid_type: (string) model horizontal grid type.
+    dlat: (float) meridional spacing between latitude grid points.
+    """
+    min_latitude = latitude_points[0]
+    min_lat_v_nd_grid = -90.+0.5*dlat
+    min_lat_v_eg_grid = -90
+
+    if grid_type == GRID_END_GAME:
+        return min_latitude == min_lat_v_eg_grid
+    elif grid_type == GRID_NEW_DYNAMICS:
+        return np.allclose(min_lat_v_nd_grid, min_latitude)
+    else:
+        return False  # TODO: Is this situation ever valid?
```

Comment:
It's possible code could reach this point if a data file is supplied with a different staggering, although `get_grid_type()` should raise an exception to prevent this.

There's no harm in leaving it in as `False` is probably a more useful return value than `None`.